project-codeflare/codeflare-sdk
# Refactoring to Use Poetry

# Issue link
#274 

# What changes have been made
 - The .gitignore has been updated so that the poetry.lock file is committed to the repo - allowing everyone to use the same versions of dependencies.
 - The main requirements.txt file in the root of the repo has been removed.
 - The readme has been updated to include `poetry install` rather than using pip
     - Instructions have also been added allowing creation of a requirements.txt file if required

 - The pre-commit Containerfile has been updated to use poetry, the build command will now be run from the root directory to allow access to the pyproject.toml file and all required dependencies.

 - Any installation commands and documentation/readmes have been updated to reflect the changes.

 - We will not remove the requirements.txt files from the demo notebooks. For more info see thread [here](https://project-codeflare.slack.com/archives/C04LK0G441F/p1692108534524669?thread_ts=1692010942.125799&cid=C04LK0G441F).

# Verification steps
- First checkout this pr
- Run `poetry install`
- Run `poetry build`

## To verify the building of the precommit image
 - In the root directory run `podman build -f .github/build/Containerfile .` and ensure that the build completes successfully.

## Ensure that unit tests pass
 - In the root directory run the unit tests with `poetry run pytest -v tests/unit_test.py`  

## Test demo notebooks
 - Create a cluster
 - Clone the distributed workloads repo
 - Log in to your cluster
 - Install the codeflare stack from the distributed workloads repo by running the following command `make all-in-one`
 - Once the installation of all components is complete, navigate to the ODH dashboard and to a Jupyter Notebook, and navigate to the terminal.
 - Uninstall the current codeflare-sdk by running  `pip uninstall codeflare-sdk -y`
 - Drag and drop the repo build created by the poetry build command run earlier into the Jupyter notebook and run `pip install codeflare_sdk-0.0.0.dev0-py3-none-any.whl`
- Restart the Notebook kernel
- Run through a couple of demo notebooks and ensure that they run successfully



## Checks
- Testing Strategy
